elyra/validation/validate.py

```diff
@@ -0,0 +1,424 @@
+        """
+        label_name_max_length = 63
+
+        if len(label_name) > label_name_max_length:
+            response.add_message(severity='1',
+                                 message_type="invalidNodeLabel",
+                                 message="Property string value has exceeded the max length allowed ",
+                                 data={"nodeID": node_id,
+                                       "propertyName": 'label',
+                                       "value": label_name})
+
+        # TODO: run regex check on label
+
+    def validate_pipeline_graph(self, pipeline: Dict, response):
+        """
+        Validates that the pipeline is an acyclic graph, meaning no circular references
+        :param response: ValidationResponse containing the issue list to be updated
+        :param pipeline: A dictionary describing the pipeline
+        """
+        # TODO: Supernodes?
+
+        pipeline_json = json.loads(json.dumps(pipeline))
+
+        # List of nodes in the pipeline
+        node_list = pipeline_json['pipelines'][0]['nodes']
+        pipeline_id = pipeline_json['pipelines'][0]['id']
+
+        # Build the list[tuple] of edges in the pipeline
+        node_graph = []
+        for node in node_list:
+            if 'links' in node['inputs'][0]:
+                for link in node['inputs'][0]['links']:
+                    node_graph.append((node['id'], link['node_id_ref']))
+            else:
+                node_graph.append((node['id'], ""))
+
+        graph = nx.DiGraph(node_graph)
+
+        for node in node_list:
+            graph.add_node(node['id'])
+
+        # gets any node that is not connected to the pipeline
+        for isolate in nx.isolates(graph):
+            response.add_message(severity='2',
+                                 message_type="singletonReference",
+                                 message="This node is not connected to any part of the pipeline",
+                                 data={"nodeID": isolate,
+                                       "pipelineID": pipeline_id})
+
+        # get all cycles detected in the graph
+        for cycle in nx.simple_cycles(graph):
+            if len(cycle) == 1:
+                for node in node_list:
+                    for link in node['inputs'][0]['links']:
+                        if cycle[0] == link['node_id_ref']:
+                            response.add_message(severity='1',
+                                                 message_type="circularReference",
+                                                 message="A cycle was found within this pipeline",
+                                                 data={"nodeID": node['id'],
+                                                       "pipelineID": pipeline_id})
+            else:
+                for node in node_list:
+                    for link in node['inputs'][0]['links']:
+                        if cycle[0] == link['node_id_ref'] and cycle[1] == node['id']:
```

Comment:
Are there at most two cycles?

Reply:
nope there could be more than two. 
further above I iterate thru the cycles found in the graph. 
for each cycle I check to see if its a single node ID, if so, I know its a cycle into itself e.g. node1 -> node1.....
otherwise it should return an edge representing the first edge of the cycle...e.g. node1->node2 

this else statement is just a look-up loop to find the link ID associated with that edge.

Reply:
Would you mind adding comments as to how cycles are detected?

Reply:
sure no problem, its mostly the networkx lib doing the work though

Reply:
shoot, welp its back to the drawing board, as the function call to look for cycles does not do a complete np depth first lookup and only returns the first cycle. will tackle this on monday

Reply:
Would the `simple_cycles()` method be better for this?  Do we care if there are multiple cycles?  Yes, it would be great to be able to highlight various cycles when multiple are present, but I suspect pointing out one (the first), would alert the user that multiple could be in play and they would either fix the others or find out the next when validating again.

Regarding the code, I have a few more comments:
1. It seems like it would be easier (more clear) to start from an empty graph and, in a single pass through the node_list, add both the node and its edges - avoiding the second loop to add the nodes.
2. I think the edge coordinates might be reversed.  Since these are _inputs_ to the current node, the edge data seems like it should be the `node_id_ref` first, then _this_ node id since the edge is input _to_ this node.
3. I'm not sure we need the specification of the empty string if a given node has no inputs - that doesn't seem correct.  The isolated nodes should be identifiable by not appearing in any edge tuples.

This would yield something like the following...
```python
        graph = nx.DiGraph()
        for node in node_list:
            graph.add_node(node['id'])
            if 'links' in node['inputs'][0]:
                for link in node['inputs'][0]['links']:
                    graph.add_edge((link['node_id_ref'], node['id']))
```

Regarding cycles, I think we do want to use `simple_cycles()` which returns a list of cycles where each cycle is a list of edges (tuples).
Since the graph is built from the _inputs_, the "interesting node" would be the second element of the edge tuple.  It seems like, for cycles, the `data` within the `ValidationResponse` should simply be a list of nodes composing the simple cycle - which can be built directly from the results of the simple cycle call and we shouldn't need to walk the nodes, but just use the second tuple of each cycle element's edge.

```python
        cycles_detected = nx.simple_cycles(graph)

        for cycle in cycles_detected:
            nodes_in_cycle = []
            for edge_u, edge_v in cycle:
                nodes_in_cycle.append(edge_v)
                
            response.add_message(severity=ValidationSeverity.Error,
                                                      message_type="circularReference",
                                                      message="A cycle was found within this pipeline",
                                                      data={"pipelineID": pipeline_id, "nodes_in_cycle": nodes_in_cycle})
```
I don't know what the consumer of these response objects expects in this case, so the array may need to be split up across multiple messages - one for each node in a given cycle - and moved to within the inner loop, eliminating the need for the `nodes_in_cycles` array.